### PR TITLE
TEP17: Implement plots with pyqtgraph

### PR DESCRIPTION
**Pull request by cpascual**
This Pull request aims to introduce the TEP17 (see its main text [here](https://github.com/cpascual/taurus/blob/tep17/doc/source/tep/TEP17.md)) :

    Title: Implement plots with pyqtgraph
    Abstract: 
     Replace TaurusPlot and TaurusTrend implementations from the deprecated
     taurus.qt.qtgui.qwt5 by taurus_pyqtgraph implementations

### CODE REVIEW

**Comment by cpascual** [comment]
Hi,

Let me give some context for this TEP:

The necessity for a replacement for PyQwt-based widgets was already acknowledged some years ago. Since then, I have been keeping an eye on the possible alternatives (mostly guiqwt, PythonQwt, PyMca/Silx), but I lacked the time to attempt the implementation.
The situation has been finally unblocked when @oscarPrades started a 6-months part-time internship at ALBA focused on writing the needed replacement (this TEP DRAFT already includes proof-of concept implementation of some of the required features)

Please  comment and discuss here whether you think that the general orientation is acceptable (in which case we would pass this TEP from DRAFT to CANDIDATE and continue with the implementation of the rest of the features)

File: doc/source/tep/TEP17.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+  - Support for limiting curve buffers (C8)
+  - UI for curve buffers (C2)
+
+
+### For 2D plots (images)
+
+  - (*) Plot a single image (M0)
+  - (*) UI for Add/remove image (M4)
+  - Cross sections (slicing) (S4)
+  - Slicing (S4)
```
Comment (LJBD) [inline doc/source/tep/TEP17.md]: Is this repetition of slicing intentional? If so, could you please clarify what you mean in each of those points?

**Comment by LJBD** [comment]
I like the idea very much!

File: doc/source/tep/TEP17.md (hunk 1 above)
Comment (cpascual) [reply doc/source/tep/TEP17.md]: no. it was a copy-paste bug. fixed

**Comment by cpascual** [comment]
Let's do some summary. Here is a task list so that we can get an idea on how the implementation is progressing. I will put here only the features that are marked to be part of the initial implementation (marked with "*" in the TEP17 document). In this task list we check the tasks for which there is already a prototype ready:

### For 1D plots/trends

- [x] 1D plot: plot of multiple 1D models with auto-changing color and
    availability of legend (M8)
- [x] Date-time support on X axis (display only, see "UI for
    setting scale limits *in date/time format*" below) (M16)
- [ ] Stand-alone widget (M2)
- [x] Zooming & panning with "restore original view" option (not the same
    as zoom stacking, see below) (M0)
- [x] Possibility to use (at least) 2 Y-scales (M12)
- [x] UI for adding taurus curves via ModelChooser. See also
    "Improved Model Chooser" below  (M4)
- [ ] Store/retreive configuration (save/load settings) (M8)
- [x] Support for non-taurus curves in same plot (aka "raw data") (S0)
- [x] UI for setting scale limits and lin/log options (S0)
- [x] Export data as ascii: without date-time support (S0)
- [x] Export plot as image (S0)
- [x] UI for moving a curve from one Y-scale to another (S12)
- [x] UI for choosing line color, thickness symbol, filling... (S16)
- [ ] Arbitrary Label scale (aka FixedLabelsScale) (C8)

### For 1D trends

Most of the features mentioned for 1D plots affect the 1D trends as
well. Apart from those, here is a list of more specific features of
trends:

- [x] "1D trends": plot of scalars vs event number or timestamp (M16)
- [ ] Fixed-range scale (aka oscilloscope mode) (M8)
- [ ] UI to switch between fixed and free scale mode (S12)
- [ ] Support for forced-reading of attributes (aka "-r mode") (M10)
- [ ] UI for forced-reading mode (C2)


### For 2D plots (images)

- [x] Plot a single image (M0)
- [x] UI for Add/remove image (M4)
- [ ] "calibrated" XYImage (assigning values to X and Y scale, as in
    guiqwt's XYImageItem) S8


### For 2D trends (spectrograms)

Most of the features for 2D plots affect also the 2D trends. Apart
from those, here is a list of more specific features of 2D trends:

- [ ] Absolute date-time scale (display, see same feat in TaurusPlot)
- [ ] Fixed-range scale (aka oscilloscope mode, same as for 1Dtrends) (M8)
- [ ] UI to switch between fixed and free scale mode (S12)

**Comment by cpascual** [comment]
Note: I am moving this development to an independent repo providing the taurus_pyqtgraph plugin (which thanks to #684 will be loaded transparently as taurus.qt.qtgui.tpg if installed). In this way, the develpment of this TEP will be decoupled from the release schedule of Taurus.

**Comment by cpascual** [comment]
Since taurus_pyqtgraph 0.4.9 already checks all the marks for the TEP17, I think it is time to finally propose TEP17 for APPROVAL

This is a call to @taurus-org/integrators to cast a vote on this.


Note that this is mostly a formality because this PR in itself only adds the TEP text, and the actual implementation is already done and used in the  [taurus_pyqtgraph](https://github.com/taurus-org/taurus_pyqtgraph) plugin.

**Review by teresanunez — APPROVED** [review]

**Review by reszelaz — APPROVED** [review]
I've re-read the TEP again and it looks ok to me to ACCEPT it.

**Comment by cpascual** [comment]
Hello @HEnquist , I am told that you may be the person appointed at MAXIV for voting on  #452 and #1168

If so, would you mind adding a revision (or simply a comment) to these PRs stating the MAXIV vote?

**Comment by cpascual** [comment]
Ok, I got ok revisions from @reszelaz  and @teresanunez , so I am going to promote TEP17 to accepted state.

(finally!)

**Comment by HEnquist** [comment]
Very sorry for never replying! But yes we agree.

**Pull request merged**